core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigDefinition.java

```diff
@@ -264,59 +274,113 @@ private GroupConfigType processConfigGroup(final String containingName, final Co
                 final LeafConfigType leaf;
                 // optional config property
-                gct.addField(leaf = new OptionalObjectConfigType(field.getName(), gct, consume,
-                        defaultValue.equals(ConfigItem.NO_DEFAULT) ? "" : defaultValue, rawTypeOfParameter(fieldType, 0),
-                        javadocKey, subKey));
+                final Class<?> optionalType = rawTypeOfParameter(fieldType, 0);
+                OptionalObjectConfigType optionalObjectConfigType = new OptionalObjectConfigType(field.getName(), gct, consume,
+                        defaultValue.equals(ConfigItem.NO_DEFAULT) ? "" : defaultValue, optionalType,
+                        javadocKey, subKey, loadEnhancedConverter(field, optionalType, subKey));
+                gct.addField(leaf = optionalObjectConfigType);
                 container.getConfigDefinition().getLeafPatterns().addPattern(subKey, leaf);
             } else {
                 final LeafConfigType leaf;
                 // it's a plain config property
-                gct.addField(leaf = new ObjectConfigType(field.getName(), gct, consume,
-                        mapDefaultValue(defaultValue, fieldClass), fieldClass, javadocKey, subKey));
+                ObjectConfigType objectConfigType = new ObjectConfigType(field.getName(), gct, consume,
+                        mapDefaultValue(defaultValue, fieldClass), fieldClass, javadocKey, subKey,
+                        loadEnhancedConverter(field, fieldClass, subKey));
+                gct.addField(leaf = objectConfigType);
                 container.getConfigDefinition().getLeafPatterns().addPattern(subKey, leaf);
             }
         }
         return gct;
     }
 
+    private Class<? extends Converter<?>> loadEnhancedConverter(Field field, Class<?> clazz, String configProperty) {
+        final DefaultConverter defaultConverter = field.getAnnotation(DefaultConverter.class);
+        final ConvertWith convertWith = field.getAnnotation(ConvertWith.class);
+
+        if (defaultConverter != null && convertWith != null) {
+            throw new IllegalArgumentException(String.format(
+                    "Duplicate conversion behaviour specified on property %s : %s annotation and %s annotation given",
+                    configProperty, DefaultConverter.class.getName(), ConvertWith.class.getName()));
+        }
+
+        if (defaultConverter != null) {
+            return null; // use implicit converters or custom converters
+        }
+
+        if (convertWith != null) {
+            if (clazz.isEnum()) {
+                return HyphenateEnumConverter.class;
+            } else {
+                Class<? extends Converter<?>> converterClass = convertWith.value();
+                try {
+                    final Method method = converterClass.getMethod("convert", String.class);
+                    final Type type = method.getAnnotatedReturnType().getType();
+                    if (clazz.equals(type)) {
```

Comment:
This should be `class.isAssignableFrom(getRawTypeOf(type))` I would think, because the converter output type can be more specific than the accepting member's type (for example I could use a `Converter<String>` to assign a value into a field of type `Object`).